File: AccuPay/Views/Payroll Forms/Loans/LoanRowRecord.vb
Option Strict On

Imports AccuPay.Core.Entities
Imports AccuPay.Core.Interfaces.Excel
Imports AccuPay.Utilities.Attributes

Public Class LoanRowRecord
    Implements IExcelRowRecord

    <ColumnName("Employee Name")>
    Public Property EmployeeFullName As String

    <ColumnName("Employee ID")>
    Public Property EmployeeNumber As String

    Public Property LoanName As String

    <ColumnName("Loan number/code")>
    Public Property LoanNumber As String

    <ColumnName("Start date")>
    Public Property StartDate As Date?

    Public Property TotalLoanAmount As Decimal?

    <ColumnName("Loan balance")>
    Public Property TotalBalanceLeft As Decimal?

    Public Property DeductionAmount As Decimal?

    <ColumnName("Deduction frequency(First half, End of the month, Per pay period)")>
    Public Property DeductionSchedule As String

    Public Property Comments As String

    <Ignore>
    Public Property ErrorMessage As String

    <Ignore>
    Public Property LineNumber As Integer Implements IExcelRowRecord.LineNumber

    <Ignore>
    Public Property LoanType As Product

    Public Function ToLoan(employeeId As Integer) As Loan

        Dim newLoan = New Loan With {
            .RowID = Nothing,
            .OrganizationID = z_OrganizationID,
            .EmployeeID = employeeId,
            .LoanNumber = LoanNumber,
            .Comments = Comments,
            .TotalLoanAmount = TotalLoanAmount.Value,
            .TotalBalanceLeft = TotalBalanceLeft.Value,
            .DedEffectiveDateFrom = StartDate.Value,
            .DeductionAmount = DeductionAmount.Value,
            .DeductionPercentage = 0,
            .LoanName = LoanName,
            .LoanTypeID = LoanType.RowID,
            .Status = Loan.STATUS_IN_PROGRESS,
            .DeductionSchedule = DeductionSchedule
        }

        newLoan.RecomputeTotalPayPeriod()
        newLoan.RecomputePayPeriodLeft()

        Return newLoan

    End Function

End Class

File: AccuPay.Core/Services/Imports/Loans/LoanRowRecord.cs
using AccuPay.Core.Interfaces.Excel;
using AccuPay.Utilities.Attributes;
using System;
using System.ComponentModel.DataAnnotations;

namespace AccuPay.Core.Services.Imports.Loans
{
    public class LoanRowRecord : ExcelEmployeeRowRecord, IExcelRowRecord
    {
        [Required]
        [ColumnName("Employee ID")]
        public string EmployeeNo { get; set; }

        [Required]
        [ColumnName("Loan name")]
        public string LoanName { get; set; }

        [ColumnName("Loan number/code")]
        public string LoanNumber { get; set; }

        [Required]
        [ColumnName("Start date")]
        [DataType(DataType.Date, ErrorMessage = "Invalid Date Value")]
        public DateTime? StartDate { get; set; }

        [Required]
        [ColumnName("Total loan amount")]
        [Range(0, double.MaxValue, ErrorMessage = "Invalid Total loan amount")]
        public decimal? TotalLoanAmount { get; set; }

        [Required]
        [ColumnName("Loan balance")]
        [Range(0, double.MaxValue, ErrorMessage = "Invalid Loan balance")]
        public decimal? TotalLoanBalance { get; set; }

        [Required]
        [ColumnName("Deduction amount")]
        [Range(0, double.MaxValue, ErrorMessage = "Invalid Deduction amount")]
        public decimal? DeductionAmount { get; set; }

        [Required]
        [ColumnName("Deduction frequency(First half, End of the month, Per pay period)")]
        public string DeductionSchedule { get; set; }

        [ColumnName("Comments")]
        public string Comments { get; set; }

        public int LineNumber { get; set; }
    }
}
